ingest/expression_files/dense_ingestor.py

```diff
@@ -40,21 +40,36 @@ def __init__(self, file_path, study_file_id, study_id, **matrix_kwargs):
         self.gene_names = {}
         self.header = DenseIngestor.process_header(next(self.csv_file_handler))
 
+    def matches_file_type(file_type):
+        return file_type == "dense"
+
     def execute_ingest(self):
         # Method can only be executed once due to
         # dependency on position in text file.
         # Row after header is needed for R format validation
         row = next(self.csv_file_handler)
-        if not DenseIngestor.is_valid_format(self.header, row):
+        if not DenseIngestor.is_valid(
+            self.header,
+            row,
+            query_params=(self.study_id, self.mongo_connection._client),
+        ):
             raise ValueError("Dense matrix has invalid format")
         # Reset csv reader to first gene row
         self.csv_file_handler = self.open_file(self.file_path)[0]
         next(self.csv_file_handler)
         for gene_docs, data_array_documents in self.transform():
             self.load(gene_docs, data_array_documents)
 
-    def matches_file_type(file_type):
-        return file_type == "dense"
+    @staticmethod
+    def is_valid(header, row, query_params):
+        return all(
+            [
+                DenseIngestor.has_unique_header(header),
+                DenseIngestor.has_gene_keyword(header, row),
+                DenseIngestor.header_has_valid_values(header),
+                GeneExpression.has_unique_cells(header, *query_params),
```

Comment:
this method needs a way to return the specific problem -- I recommend changing the method so it returns a tuple of [boolean, string], where the boolean is true/false for valid /invalid, and the string is the error message if invalid

Reply:
Tracked [here](https://broadworkbench.atlassian.net/browse/SCP-2660)